Create the English basics/functions lesson (#24)
haskellfoundation/HaskellSchool: en/lessons/basics/functions.md

---
version: 1.0.0
title: Functions
---

> Functions are at the heart of every Haskell program. Let us now see how to use them.

Reviewer: I think "Let us now see how to use them" lacks coherence / doesn't introduce new information to the reader. Perhaps you could talk about the nature of the article (e.g. "You'll see some basic examples of functions below").

{% include toc.html %}

## Anonymous Functions

> Lambdas are one of the building blocks of Haskell, and their syntax is the following:

Reviewer: ```suggestion
Lambdas are one of the building blocks of Haskell, and they have the following syntax:
```

I think having the adjective before the noun sounds more natural here.

```
   Structure of a lambda

   \x -> x
    │ │  └─ Body of the lambda
    │ └─ Separator between head and body of the lambda
    └─ Declaration of the input bindigs
```

> And incidentally, we have also seen one of the most primordial functions : The identity function, or `id`.  

Reviewer: `:` shouldn't have a space before it, and perhaps "you have just seen" would better highlight that you're referring to `\x -> x`. 

I'm also not sure about the usage of "we" here (and in general): I haven't needed to use it in `basics.md`, but if I were to use it, I'd use it to refer to the general Haskell userbase or the writers of the article (e.g. "We introduced").

Author: Ok!

Reviewer: I think that "The" shouldn't be capitalised here.

This function is very simple, as it returns precisely the argument that we pass to it.

## Named functions

However, one can only go so far with anonymous functions. This is why we can give names to our functions:

```console?lang=haskell&prompt=ghci>,ghci|
ghci> addOne x = x + 1
ghci> addOne 2
3
```

## Higher-order functions

> We can now see a third way of using functions, which is the composition of both anonymous and named functions:

Reviewer: Although higher-order functions can compose anonymous and named functions, I'm not sure it's the best way of defining them.

Reviewer: I think what I wanted is said more clearly here: https://github.com/haskellfoundation/HaskellSchool/pull/24#issuecomment-873465708.

> Higher-order functions!

Reviewer: In British English, even if a colon is succeeded by a full sentence, you don't capitalise the succeeding clause.

Author: Ah, didn't know that. :) Thanks!

They become particularly useful when you need to control the application of a function over a parameter:

```console?lang=haskell&prompt=ghci>,ghci|

> ghci> apply2 f x = f (f x)

Reviewer: Perhaps this function should be called `applyTwice`. I think it would be more consistent with the other function names (like `addOne`) in this document.

Author: yep

ghci> apply2 addOne 2
4
```

## Local definitions: `where` & `let … in`

You can introduce bindings, such as local definition or intermediate results, through two different syntaxes:  

> The `where` keyword, and the `let … in` construct.

Reviewer: Is this line supposed to be rendered on a new line? It seems that the line breaks early owing to the newline in the file.

Author: Yeah the extra `  ` at the end of the previous line is for a newline

### `where`

Here is an example of a function using `where` to store local functions:

```haskell
processInt x = timesTwo (addOne x)
  where
    addOne y = y + 1
    timesTwo z = z * 2
```

Other functions will be unable to use `addOne` and `timesTwo` since they belong to the scope of `processInt`.

We can use its sibling, `let`, to produce the same result:

```haskell
processInt2 x =
  let addOne y = y + 1
      timesTwo z = z * 2  
   in timesTwo (addOne x)
```

> Those two syntaxes exist to better let you write down your mental process. Even though it is good form to stick with one

Reviewer: I believe it should be `These`, because you introduced the concept.

when in a collaborative setting (at work, for example), you are absolutely free to pick the one that makes you happy to
write Haskell. :)

---

> And that is for the basics about functions! In the next lesson, you will see that functions not only have a body, but also a type.

Reviewer: ```suggestion
And that is it for the basics about functions! In the next lesson, you will see that functions not only have a body, but also a type.
```

Author: Thanks!

title: Functions
---

> Functions are at the heart of every Haskell program, and are first-class in the language.

Reviewer 2: maybe elaborate what first-class means

You'll see some examples of functions below.

{% include toc.html %}

## Anonymous Functions

> Anonymos functions, or Lambdas, are one of the building blocks of Haskell. They have the following syntax:

Reviewer 2: Anonymos -> Anonymous

**Structure of a lambda**
```

>  ╭─ "head", the declaration of the input bindigs

Reviewer 2: bindigs -> bindings, but I would suggest to avoid calling it the "head" and just call this "the input of the function"

 │ 
 │ ╭─ Separator between head and body
 │ │ 

>  │ │  ╭─ Body of the lambda

Reviewer 2: I would call this "the output of a function"

\x -> x
```

> And incidentally, you have just seen one of the most primordial functions: The identity function, or `id`.  

Reviewer 2: maybe instead of "just seen" use "the example above is"?
I would suggest elaborating a bit more about the structure of a lambda: the input parameter can be any name, and the output any expression, the body can use any name that is currently in scope, including the parameter. for example: 

```hs
let
  greeting = "Welcome"
in
  \name -> greeting ++ " " ++ name
```

Or something similar. Maybe also mention closures at the same time? This could be further elaborated in the next section about currying (which is basically creating closures).

### Syntactic Sugars

#### Currying

In Haskell, all your functions have one argument. Inspired by the Lambda Calculus, this design is however extended in
the language through syntactic sugar that allows you to express yourself better:

```haskell
-- Syntactic sugar
\x y -> x + y
 
-- Currified version
\x -> \y -> x + y
```

The term is currification: Your function of two arguments has been transformed behind the scenes into a function
of one argument containing another function of one argument, containing itself the operation.  
In Javascript, this would be expressed like this:

```javascript
// Syntactic sugar
function (x, y) {
  return x + y;
}

// Currified version
function (x) {
  return function (y) {
    return x + y;
  };
}

> let a = function (x) { return function (y) { return x + y }; }
> a(1)
[Function (anonymous)]
> a(1)(2)
3
```

#### Tacit, or "point-free" style

Tacit style allows you to omit the arguments (or "points") of a function when declaring or applying it.  
While it can be quicker and help with equational reasoning, the loss of readability that it incurs needs to be a factor
in your decision to use it.

> For example, the function that adds 1 to a number, `(\x -> x + 1)` can be reduced to `(+1)`.  

Reviewer 2: Consider using a function instead of an operator here, interesting candidates are `elem`, `min`, `replicate`, etc. and maybe also mention that we can only replace the last argument, so `\n -> elem n [1, 2, 3]` can't be reduced to `elem [1, 2, 3]`.

These functions are equivalent in every point except their syntax.

### Function Application

The function application syntax is the following:

```console?lang=haskell&prompt=ghci>,ghci|
ghci> (\x y -> x + y) 2 3 
5
```

> Here, the function between parentheses has been fully applied. But remember that these

Reviewer 2: Maybe explain what function application means (supplying values as parameters to functions), and fully applied then means "the function was supplied with all of it's arguments). Explicitly mention that supplying a value means that we remove the first layer of lambda (the first input argument) and replace (or substitute) the name with the supplied value in the body. Maybe also mention that this is formally called substitution?

are actually two functions on top of each-other, which means that,
much like in the above JavaScript example, it will return a function if you only pass it one argument:

```
(\x y -> x + y) 2
│ 
╰─ (\y -> 2 + y) 5
   │ 
   ╰─ 2 + 5
```

This ability to do partial application unlocks many abilities, one of which allowing you to build computations whose
parameters are not available all at the same time during the program's lifetime.

## Named functions

> However, one can only go so far with anonymous functions. This is why you can give names to your functions:

Reviewer 2: Maybe this is a good place to talk about what first-class means, lambdas can be used just like normal data values such as Int and Bool, they can be passed to functions and returned from functions as we've seen, can be put in a data structure such as a list, and they can be *given a name* (`increment = \n -> n + 1`). There's even a syntactic sugar for the latter:

```console?lang=haskell&prompt=ghci>,ghci|
ghci> addOne x = x + 1
ghci> addOne 2
3
```

## Higher-order functions

> There exists a third kind of functions, whose first argument is itself an anonymous functions: higher-order functions.

Reviewer 2: Maybe instead of saying that these are a third kind, just mention that this is a fancy name for functions that take other functions as input? It's not a different feature from what we've seen (as you write below), just a useful pattern that is useful in many contexts. And it doesn't have to be the first argument.

> Because functions are first-class components of Haskell, they are no more magical than anonymous or named functions.

Reviewer 2: :+1:

They become particularly useful when you need to control the application of a function over a parameter:

```console?lang=haskell&prompt=ghci>,ghci|
ghci> applyTwice f x = f (f x)
ghci> applyTwice addOne 2
4
```

Some of the most well-known higher-order functions out there are `map` and `filter`

```console?lang=haskell&prompt=ghci>,ghci|
ghci> map (\x -> x + 2) [1,2,3,4]
[3,4,5,6]

> ghci> filter (\x -> x > 3) [1,2,3,4,5,6]

Reviewer 2: maybe use `even` or `odd` for an example with regular functions? :)

[4,5,6]
```

> ## Local definitions: `where` & `let … in`

Reviewer 2: In my opinion there should be an earlier section about naming things where this should exist, should also cover indentation (how do I know where a definition ends?)

You can introduce bindings, such as local definition or intermediate results, through two different syntaxes:  
The `where` keyword, and the `let … in` construct.

### `where`

Here is an example of a function using `where` to store local functions:

```haskell
processInt x = timesTwo (addOne x)
  where
    addOne y = y + 1
    timesTwo z = z * 2
```

Other functions will be unable to use `addOne` and `timesTwo` since they belong to the scope of `processInt`.

### `let … in`

you can use its sibling, `let`, to produce the same result:

```haskell
processInt2 x =
  let addOne y = y + 1
      timesTwo z = z * 2  
   in timesTwo (addOne x)
```

These two syntaxes exist to better let you write down your mental process. Even though it is good form to stick with one
when in a collaborative setting (at work, for example), you are absolutely free to pick the one that makes you happy to
write Haskell. :)

> ## Function Composition

Reviewer 2: I like this section and think it is very important, I'm not a huge fan of introducing mathematical notation though, but up to you.

Author: I don't have a better way to show that you have a way to go from X to Y, and then from Y to Z, and their composition gives us X to Z :(

Author: But we can go back on this later.

Reviewer 2: Yeah you can just merge it and go back to it after the tutorial is pretty much complete and figure out if you want to change it

Function composition is one of the most primordial aspects of functional programming, and Haskell offers builtin operators
to enable it.

The `(.)` operator that is used for function composition is a legacy of the mathematical notation:

Let two functions  
```
f : X |-> Y  
g : Y |-> Z  
```

To go from X to Z, these two functions are composed like this:

`g(f(x))` -- read "g of f"

To simplify the notation, this composition is also written:

`g ∘ f : X |-> Z`

Which can be applied to an argument like this:

`(g ∘ f)(x)`

And this notation is so convenient that Haskell has kept it.

As such, if you have two or more functions that can fit like Lego bricks, nothing more is needed than simply getting

> the arguments right. Remember the `processInt` function from earlier? It can now be written like this:

Reviewer 2: Consider using the `map` and `filter` examples above for the example :)

title: Functions
---

Functions are at the heart of every Haskell programme, and are first-class in the language.  
They can be passed as parameters to other functions, and returned as computation results.  
You'll see some examples of functions below.

{% include toc.html %}

## Anonymous Functions

Anonymous functions, or Lambdas, are one of the building blocks of Haskell. They have the following syntax:

**Structure of a lambda**
```
 ╭─ input of the function
 │ 
 │ ╭─ Separator between input and output
 │ │ 
 │ │  ╭─ output of the function
\x -> x
```

And incidentally, the example above is one of the most primordial functions: The identity function, or `id`.  
This function is very simple, as it returns precisely the argument that you pass to it.

> In Lambdas, the input parameter can be any name, the output can be any expression.

Reviewer: ```suggestion
In Lambdas, the input parameter can be any name, and the output can be any expression.
```
This removes a comma splice.

### Syntactic Sugar: Currying

In Haskell, all your functions have one argument. Inspired by the Lambda Calculus, this design is however extended in
the language through syntactic sugar that allows you to express yourself better:

```haskell
-- Syntactic sugar
\x y -> x + y
 
-- Currified version
\x -> \y -> x + y
```

> The term is currification: Your function of two arguments has been transformed behind the scenes into a function

Reviewer: ```suggestion
The term is currification: your function of two arguments has been transformed behind the scenes into a function
```
I believe British English has slightly different capitalisation rules.

Author: Yep, only proper nouns and acronyms are capitalised after a colon in en_UK

╰─ (\y -> 2 + y) 5
   │ 
   ╰─ 2 + 5
```

This ability to do partial application unlocks many abilities, one of which allowing you to build computations whose
parameters are not available all at the same time during the programme's lifetime.

## Named functions

However, one can only go so far with anonymous functions. This is why you can give names to your functions:

```console?lang=haskell&prompt=ghci>,ghci|
ghci> addOne x = x + 1
ghci> addOne 2
3
```

## Higher-order functions

There exists a third kind of functions, whose first argument is itself an anonymous functions: higher-order functions.
Because functions are first-class components of Haskell, they are no more magical than anonymous or named functions.

They become particularly useful when you need to control the application of a function over a parameter:

```console?lang=haskell&prompt=ghci>,ghci|
ghci> applyTwice f x = f (f x)
ghci> applyTwice addOne 2
4
```

Some of the most well-known higher-order functions out there are `map` and `filter`

```console?lang=haskell&prompt=ghci>,ghci|
ghci> map (\x -> x + 2) [1,2,3,4]
[3,4,5,6]

ghci> filter odd [1,2,3,4,5,6]
[1,3,5]
```

## Local definitions: `where` & `let … in`

> You can introduce bindings, such as local definition or intermediate results, through two different syntaxes:  

Reviewer: ```suggestion
You can introduce bindings, such as local definitions or intermediate results, through two different syntaxes:  
```

The `where` keyword, and the `let … in` construct.

### `where`

Here is an example of a function using `where` to store local functions:

```haskell
processInt x = timesTwo (addOne x)
  where
    addOne y = y + 1
    timesTwo z = z * 2

> ```

Reviewer: Could you possibly show that the variable `x` can be used within the bindings?

Author: damnit, yes

Other functions will be unable to use `addOne` and `timesTwo` since they belong to the scope of `processInt`.

### `let … in`

> you can use its sibling, `let`, to produce the same result:

Reviewer: ```suggestion
You can use its sibling, `let`, to produce the same result:
```

title: Functions
---

Functions are at the heart of every Haskell program, and are first-class in the language.  
They can be passed as parameters to other functions, and returned as computation results.  
You'll see some examples of functions below.

{% include toc.html %}

## Anonymous Functions

Anonymous functions, or Lambdas, are one of the building blocks of Haskell. They have the following syntax:

**Structure of a lambda**
```
 ╭─ input of the function
 │ 
 │ ╭─ Separator between input and output
 │ │ 

>  │ │  ╭─ output of the function

Reviewer: Could all of the annotations have their first letter capitalised?

\x y -> x + y
 
-- Currified version
\x -> \y -> x + y
```

The term is currification: your function of two arguments has been transformed behind the scenes into a function
of one argument containing another function of one argument, containing itself the operation.  
In Javascript, this would be expressed like this:

```javascript
// Syntactic sugar
function (x, y) {
  return x + y;
}

// Currified version
function (x) {
  return function (y) {
    return x + y;
  };
}

> let a = function (x) { return function (y) { return x + y }; }
> a(1)
[Function (anonymous)]
> a(1)(2)
3
```

### Function Application

The function application syntax is the following:

```console?lang=haskell&prompt=ghci>,ghci|
ghci> (\x y -> x + y) 2 3 
5
```

Here, the function between parentheses has been fully applied. But remember that these
are actually two functions on top of each-other, which means that,
much like in the above JavaScript example, it will return a function if you only pass it one argument:

```
(\x y -> x + y) 2
│ 
╰─ (\y -> 2 + y) 5
   │ 
   ╰─ 2 + 5
```

This ability to do partial application unlocks many abilities, one of which allowing you to build computations whose

> parameters are not available all at the same time during the program's lifetime.

Reviewer: I feel like this would be clearer if it were split into two sentences. 

Perhaps I'm misunderstanding what is meant, but isn't this moreso a matter of convenience? If you didn't have currying, you could pass over the parameter instead of the function, although that would be clunkier (and it would make dealing with scope more of a nuisance).

Author: Yeah you'd have to do like the JS example, with nested `returns`

## Named functions

However, one can only go so far with anonymous functions. This is why you can give names to your functions:

```console?lang=haskell&prompt=ghci>,ghci|
ghci> addOne x = x + 1
ghci> addOne 2
3
```

## Higher-order functions

> There exists a third kind of functions, whose first argument is itself an anonymous functions: higher-order functions.

Reviewer: I agree with soupi's review here. 

Also, I believe "function" would be better suited than "functions" in some places.

Because functions are first-class components of Haskell, they are no more magical than anonymous or named functions.

They become particularly useful when you need to control the application of a function over a parameter:

```console?lang=haskell&prompt=ghci>,ghci|
ghci> applyTwice f x = f (f x)
ghci> applyTwice addOne 2
4
```

Some of the most well-known higher-order functions out there are `map` and `filter`

```console?lang=haskell&prompt=ghci>,ghci|
ghci> map (\x -> x + 2) [1,2,3,4]
[3,4,5,6]

ghci> filter odd [1,2,3,4,5,6]
[1,3,5]
```

## Local definitions: `where` & `let … in`

You can introduce bindings, such as local definitions or intermediate results, through two different syntaxes:  
the `where` keyword, and the `let … in` construct.

### `where`

Here is an example of a function using `where` to store local functions:

```haskell
processInt x = timesTwo (addOne x)
  where
    addOne y = x + 1
    timesTwo z = y * 2
```

> _Note: the binding `x` is reusable throught the rest of the bindings._

Reviewer: ```suggestion
_Note: the binding `x` is reusable throughout the rest of the bindings._
```
I'm also not sure whether this note follows the standard formatting (I guess we'd use the one that Elixir School does). 

I like this note 🙂

Author: No it's not formatted, just italicised.

These two syntaxes exist to better let you write down your mental process. Even though it is good form to stick with one
when in a collaborative setting (at work, for example), you are absolutely free to pick the one that makes you happy to
write Haskell. :)

## Function Composition

Function composition is one of the most primordial aspects of functional programming, and Haskell offers builtin operators
to enable it.

The `(.)` operator that is used for function composition is a legacy of the mathematical notation:

Let two functions  
```
f : X |-> Y  
g : Y |-> Z  
```

To go from X to Z, these two functions are composed like this:

`g(f(x))` -- read "g of f"

To simplify the notation, this composition is also written:

`g ∘ f : X |-> Z`

Which can be applied to an argument like this:

`(g ∘ f)(x)`

And this notation is so convenient that Haskell has kept it.

As such, if you have two or more functions that can fit like Lego bricks, nothing more is needed than simply getting
the arguments right. Remember the `processInt` function from earlier? It can now be written like this:

```console?lang=haskell&prompt=ghci>,ghci|
ghci> let addOne y = y + 1
ghci> let timesTwo z = z * 2

> ghci> let processInt3 x = (addOne . timesTwo) x

Reviewer: The usage of `let` here may confuse readers.

ghci> processInt3 2
5

-- And you can have some fun

> ghci> let processInt4 x = (addOne . timesTwo . addOne . timesTwo) x

Reviewer: Similar thing here with `let`.

## Named functions

However, one can only go so far with anonymous functions. This is why you can give names to your functions:

```console?lang=haskell&prompt=ghci>,ghci|
ghci> addOne x = x + 1
ghci> addOne 2
3
```

## Higher-order functions

Functions that take another function as their argument are called "higher-order functions".
Because functions are first-class components of Haskell, they are no more magical than anonymous or named functions.

They become particularly useful when you need to control the application of a function over a parameter:

```console?lang=haskell&prompt=ghci>,ghci|
ghci> applyTwice f x = f (f x)
ghci> applyTwice addOne 2
4
```

Some of the most well-known higher-order functions out there are `map` and `filter`

```console?lang=haskell&prompt=ghci>,ghci|
ghci> map (\x -> x + 2) [1,2,3,4]
[3,4,5,6]

ghci> filter odd [1,2,3,4,5,6]
[1,3,5]
```

## Local definitions: `where` & `let … in`

You can introduce bindings, such as local definitions or intermediate results, through two different syntaxes:  
the `where` keyword, and the `let … in` construct.

### `where`

Here is an example of a function using `where` to store local functions:

```haskell

> processInt x = timesTwo (addOne x)
>   where
>     addOne y = x + 1
>     timesTwo z = y * 2

Reviewer: ```suggestion
processInt x = timesTwo (addX x)
  where
    addX y = x + y
    timesTwo z = y * 2
```
This prevents shadowing with previous definitions and means that the variable y gets used in the function.

```

> _Note: the binding `x` is reusable throughout the rest of the bindings._

Reviewer: ```suggestion
__Note__: the binding `x` is reusable throughout the rest of the bindings.
```
This makes the formatting consistent with that of other pages (although I need to check how this formatting initially came about and it can be changed).

Author: So, we're going with bold instead of italics, right?

> Other functions will be unable to use `addOne` and `timesTwo` since they belong to the scope of `processInt`.

Reviewer: ```suggestion
Other functions will be unable to use `addX` and `timesTwo` since they belong to the scope of `processInt`.
```
This follows from previous changes.

### `let … in`

You can use its sibling, `let`, to produce the same result:

```haskell

> processInt2 x =
>   let addOne y = y + 1
>       timesTwo z = z * 2  
>    in timesTwo (addOne x)

Reviewer: ```suggestion
processInt2 x =
  let addX y = y + x
      timesTwo z = z * 2  
   in timesTwo (addX x)
```
This follows from previous changes.

```

These two syntaxes exist to better let you write down your mental process. Even though it is good form to stick with one
when in a collaborative setting (at work, for example), you are absolutely free to pick the one that makes you happy to
write Haskell. :)

## Function Composition

Function composition is one of the most primordial aspects of functional programming, and Haskell offers builtin operators
to enable it.

The `(.)` operator that is used for function composition is a legacy of the mathematical notation:

Let two functions  
```
f : X |-> Y  
g : Y |-> Z  
```

To go from X to Z, these two functions are composed like this:

`g(f(x))` -- read "g of f"

To simplify the notation, this composition is also written:

`g ∘ f : X |-> Z`

Which can be applied to an argument like this:

`(g ∘ f)(x)`

And this notation is so convenient that Haskell has kept it.

As such, if you have two or more functions that can fit like Lego bricks, nothing more is needed than simply getting

> the arguments right. Remember the `processInt` function from earlier? It can now be written like this:

Reviewer: I'm not sure that the analogy of lego bricks is clear: it may be easier to refer to two functions as being composable if the output type of one function matches the input type of the other function. A more complicated example with currying could also be given, such as `(applyTwice . (+)) 2 5`. 

Similarly, I'm not sure what "getting the arguments right" means: is this referring to the output function being valid if the order of the arguments is valid?

Reviewer: If the more complicated example were done, it would be ideal to explain it stepwise.

Author: Would this be more satisfying? 

> As such, when you have a function whose output type is the input type of another function, they are composable.
